ext/spl/spl_array.c

```diff
@@ -1790,10 +1791,19 @@ SPL_METHOD(Array, unserialize)
 		intern->ar_flags |= flags & SPL_ARRAY_CLONE_MASK;
 		zval_ptr_dtor(&intern->array);
 		ZVAL_UNDEF(&intern->array);
-		if (!php_var_unserialize(&intern->array, &p, s + buf_len, &var_hash)
-				|| (Z_TYPE(intern->array) != IS_ARRAY && Z_TYPE(intern->array) != IS_OBJECT)) {
+
+		if (!php_var_unserialize(&array, &p, s + buf_len, &var_hash)
+				|| (Z_TYPE(array) != IS_ARRAY && Z_TYPE(array) != IS_OBJECT)) {
 			goto outexcept;
 		}
+
+		if (Z_TYPE(array) == IS_ARRAY ||
+			Z_OBJ_HT(array) == &spl_handler_ArrayObject ||
+			Z_OBJ_HT(array) == &spl_handler_ArrayIterator) {
+			spl_array_set_array(object, intern, &array, 0L, 1);
+		} else {
+			ZVAL_COPY(&intern->array, &array);
+		}
```

Comment:
Two things I'm unsure about here: 1. Whether this still handles the IS_SELF case correctly (I think that might require an extra check) and 2. whether this shouldn't be using spl_array_set_array for normal objects as well. In particular that would include the check for std object properties, something that we currently don't seem to check.